### PR TITLE
fixed overflow and time conversion bugs

### PR DESCRIPTION
**Pull request by AMarcedone**

### CODE REVIEW

File: go/protocol/keybase1/extras_test.go [hunk 1]
```diff
@@ -33,6 +33,30 @@ func TestTime(t *testing.T) {
...
+	times := []time.Time{
+		time.Time{},
+		time.Now(),
+		time.Now().AddDate(1000, 0, 0),  // in a thousand years
+		time.Now().AddDate(10000, 0, 0), // in 10 thousand years
+		time.Now().AddDate(-3000, 0, 0), // 3 thousand years ago
+	}
+
+	assertEqualTimes := func(t1, t2 time.Time) {
+		require.Equal(t, t1.Unix(), t2.Unix(), "expected %v and %v to be equal", t1, t2)
```
Comment (zapu) [inline go/protocol/keybase1/extras_test.go]: can `t1.Equal` also be used to test?

Comment (zapu) [reply go/protocol/keybase1/extras_test.go]: then `time.Now()` can't be used though

File: go/protocol/keybase1/extras_test.go [hunk 2]
```diff
@@ -33,6 +33,30 @@ func TestTime(t *testing.T) {
...
 
+func TestTimeConversions(t *testing.T) {
+	times := []time.Time{
+		time.Time{},
+		time.Now(),
+		time.Now().AddDate(1000, 0, 0),  // in a thousand years
+		time.Now().AddDate(10000, 0, 0), // in 10 thousand years
+		time.Now().AddDate(-3000, 0, 0), // 3 thousand years ago
+	}
+
```
Comment (zapu) [inline go/protocol/keybase1/extras_test.go]: Can we add tests that specifically test the 0 value?
```
	require.Equal(t, UnixTime(0), ToUnixTime(time.Time{}))
	require.Equal(t, Time(0), ToTime(time.Time{}))

	require.Equal(t, time.Time{}, FromUnixTime(UnixTime(0)))
	require.Equal(t, time.Time{}, FromTime(Time(0)))
```

File: go/protocol/keybase1/extras_test.go [hunk 3]
```diff
@@ -33,6 +33,30 @@ func TestTime(t *testing.T) {
 	}
 }
 
+func TestTimeConversions(t *testing.T) {
+	times := []time.Time{
+		time.Time{},
+		time.Now(),
```
Comment (zapu) [inline go/protocol/keybase1/extras_test.go]: also test for something like
```
		time.Now().Add(time.Duration(12096123 * time.Second)),
		time.Now().Add(time.Duration(-12096321 * time.Second)),
```
both keybase1.UnixTime and keybase1.Time should be precise at least to a second

File: go/protocol/keybase1/extras_test.go (hunk 2 above)
Comment (AMarcedone) [reply go/protocol/keybase1/extras_test.go]: I think time.Time{} in the test already covers this case, but I'll put a more obvious test

Comment (zapu) [reply go/protocol/keybase1/extras_test.go]: correct me if I'm wrong, but reverse conversion test would also succeed if time.Time=0 was converted to some non integer=0  constant and then happened to convert back successfully, which is not exactly what we want

**Comment by zapu** [comment]
Thanks, I'll take another look shortly. Can we also have more eyes on this? @patrickxb @maxtaco ?

File: go/protocol/keybase1/extras_test.go (hunk 2 above)
Comment (AMarcedone) [reply go/protocol/keybase1/extras_test.go]: Oh yes, that's true. Thanks :)

File: go/protocol/keybase1/extras.go [hunk 4]
```diff
@@ -873,12 +876,11 @@ func (t Time) UnixMicroseconds() int64 {
...
 
 func ToTime(t time.Time) Time {
-	// the result of calling UnixNano on the zero Time is undefined.
-	// https://golang.org/pkg/time/#Time.UnixNano
 	if t.IsZero() {
 		return 0
 	}
-	return Time(t.UnixNano() / 1000000)
+
+	return Time(t.Unix()*1000 + (int64(t.Nanosecond()) / 1000000))
```
Comment (patrickxb) [inline go/protocol/keybase1/extras.go]: why not use `time.Unix(...)` here too?  like on line 911.

Comment (patrickxb) [reply go/protocol/keybase1/extras.go]: oh sorry, i see what this is now...

**Pull request merged**